player/loadfile.c

```diff
@@ -529,12 +529,59 @@ static bool duplicate_track(struct MPContext *mpctx, int order,
     return false;
 }
 
+static bool append_lang(size_t *nb, char ***out, char *in)
+{
+    if (!in)
+        return false;
+    char **new_out = talloc_realloc(NULL, *out, char*, (*nb + 2));
+    if (!new_out) {
+        talloc_free(in);
+        return false;
+    }
+    new_out[*nb] = in;
+    new_out[*nb + 1] = NULL;
+    *out = new_out;
+    (*nb)++;
+    return true;
+}
+
+static bool add_auto_langs(size_t *nb, char ***out)
+{
+    bool ret = false;
+    char **autos = mp_get_user_langs();
+    for (int i = 0; autos && autos[i]; i++) {
+        if (!append_lang(nb, out, talloc_steal(*out, autos[i])))
+            goto cleanup;
+    }
+    ret = true;
+
+cleanup:
+    talloc_free(autos);
+    return ret;
+}
+
+static char **process_langs(char **in)
+{
+    size_t nb = 0;
+    char **out = NULL;
+    for (int i = 0; in && in[i]; i++) {
+        if (!strcmp(in[i], "auto")) {
+            if (!add_auto_langs(&nb, &out))
+                break;
+        } else {
+            if (!append_lang(&nb, &out, talloc_strdup(out, in[i])))
+                break;
+        }
+    }
+    return out;
+}
+
 struct track *select_default_track(struct MPContext *mpctx, int order,
                                    enum stream_type type)
 {
     struct MPOpts *opts = mpctx->opts;
     int tid = opts->stream_id[order][type];
-    char **langs = opts->stream_lang[type];
+    char **langs = process_langs(opts->stream_lang[type]);
```

Comment:
do we want to do this everytime a track is chosen and not once when the option is set?

Reply:
This isn't an especially pricey operation in the scheme of things, and it's not like this is happening every frame. I wouldn't necessarily be opposed, but it's not a big enough win for me to feel like writing all the extra code to handle that.